planner/core/rule_max_min_eliminate.go

```diff
@@ -18,71 +18,188 @@ import (
-	if agg, ok := p.(*LogicalAggregation); ok {
-		// We only consider case with single max/min function.
-		if len(agg.AggFuncs) != 1 || len(agg.GroupByItems) != 0 {
-			return
+// Compose the scalar aggregations by cartesianJoin.
+func (a *maxMinEliminator) composeAggsByInnerJoin(aggs []*LogicalAggregation) (plan LogicalPlan) {
+	plan = aggs[0]
+	sctx := plan.SCtx()
+	for i := 1; i < len(aggs); i++ {
+		join := LogicalJoin{JoinType: InnerJoin}.Init(sctx)
+		join.SetChildren(plan, aggs[i])
+		join.schema = buildLogicalJoinSchema(InnerJoin, join)
+		join.cartesianJoin = true
+		plan = join
+	}
+	return
+}
+
+// checkColCanUseIndex checks the following conditions:
+// 1. whether the col is the prefix of an index.
+// 2. whether all of the selection conditions can be pushed down to the index range.
+func (a *maxMinEliminator) checkColCanUseIndex(plan LogicalPlan, col *expression.Column) bool {
+	switch p := plan.(type) {
+	case *LogicalSelection:
+		// Check whether all of the conditions can be pushed down as accessConds.
+		if _, filterConds := ranger.DetachCondsForColumn(p.ctx, p.Conditions, col); len(filterConds) != 0 {
+			return false
 		}
-		f := agg.AggFuncs[0]
-		if f.Name != ast.AggFuncMax && f.Name != ast.AggFuncMin {
-			return
+		return a.checkColCanUseIndex(p.children[0], col)
+	case *DataSource:
+		// Check whether there is an accessPath can use index for col.
+		for _, path := range p.possibleAccessPaths {
+			if path.isTablePath {
+				if col == p.handleCol {
+					return true
+				}
+			} else {
+				if col.ColName.L == path.index.Columns[0].Name.L {
+					return true
+				}
+			}
 		}
+		return false
+	default:
+		return false
+	}
+}
 
-		child := p.Children()[0]
-		ctx := p.SCtx()
-
-		// If there's no column in f.GetArgs()[0], we still need limit and read data from real table because the result should NULL if the below is empty.
-		if len(expression.ExtractColumns(f.Args[0])) > 0 {
-			// If it can be NULL, we need to filter NULL out first.
-			if !mysql.HasNotNullFlag(f.Args[0].GetType().Flag) {
-				sel := LogicalSelection{}.Init(ctx)
-				isNullFunc := expression.NewFunctionInternal(ctx, ast.IsNull, types.NewFieldType(mysql.TypeTiny), f.Args[0])
-				notNullFunc := expression.NewFunctionInternal(ctx, ast.UnaryNot, types.NewFieldType(mysql.TypeTiny), isNullFunc)
-				sel.Conditions = []expression.Expression{notNullFunc}
-				sel.SetChildren(p.Children()[0])
-				child = sel
-			}
+// cloneSubPlans clones the subPlan. We only consider `Selection` and `DataSource` here,
+// because we have restricted the subPlan in `checkColCanUseIndex`.
+func (a *maxMinEliminator) cloneSubPlans(plan LogicalPlan) LogicalPlan {
+	switch p := plan.(type) {
+	case *LogicalSelection:
+		conditions := make([]expression.Expression, 0, len(p.Conditions))
+		for _, cond := range p.Conditions {
+			conditions = append(conditions, cond.Clone())
```

Comment:
Why you clone it? What you need should be a shallow copy.

Reply:
I have considered use a shallow copy. But I'm not sure whether the latter optimization impact the condition variable.

Reply:
I prefer to use `Shallow` first as well, if we really find any problem caused by the shallow copy, we can change to `Clone` then?

Reply:
Yeah, you can make shallow copy first. I would make `Clone` unseeable in optimization phase.